File: c/euler_4.c
#include <stdio.h>
#include <string.h>
#include <math.h>
#include <stdlib.h>
#include <assert.h>

/**
 * Description: Checks if string is a palindrome
 * Params:
 * char* str - The string to check
 * int start - Index of first character
 * int end   - Index of last character
 * Returns int: 1 if string is palindrome otherwise 0
 */
int is_palindrome(char* str, int start, int end)
{
    if (end >= start && str[start] == str[end])
    {
        if (start == end || (end - start == 1 && strlen(str) % 2 == 0))
            return 1;
        else
            return is_palindrome(str, ++start, --end);
    } 
    else
    {
        return 0;
    }
}

/**
 * Description: Counts number of digits in number
 * Params: int n - The number whose digits to count
 * Returns int: number of digits
 */
int digits (int n) {
    return (int) floor(log10(abs(n))) + 1;
}

int main()
{
    int i, j, product, palindrome, result;
    i = j = product = palindrome = result = 0;

    for (i = 999; i >= 100; i--)
    {
        for (j = 999; j >= 100; j--)
        {
            product = i * j;
            char* numstr = (char*) calloc(digits(product), sizeof(char));
            assert(numstr);

            snprintf(numstr, sizeof(numstr), "%i", product);
            palindrome = is_palindrome(numstr, 0, strlen(numstr)-1);
            numstr = 0;
            free(numstr);

            if (palindrome && product > result)
                result = product;
        }
    }

    printf("%i", result);
    return 0;
}

File: javascript/euler4.js
function digit(a, b) {
    return Math.floor(a / (10**(b - 1)) % 10);
}

function isPalindrome(n) {
    const digits = Math.floor(Math.log10(n) + 1);

    for (let i = 0; i < (digits + 1) / 2; i++){
        const a = digit(n, digits-i);
        const b = digit(n, i + 1);
        if (a !== b) {
            return false;
        }
    }

    return true;
}

let result = 1;
for (let i = 100; i <= 999; i++) {
    for (let j = 100; j <= 999; j++) {
        const product = i * j;
        if (isPalindrome(product) && product > result) {
            result = product;
        }
    }
}

console.time('result');
console.log(result);
console.timeEnd('result');
